### ISSUE TITLE
[REQUEST] Temp support for older machines

### ISSUE THREAD

**Issue by RedBearAK** [aristocratos/bpytop#126]
I have a laptop that is so old it isn't properly supported by psutil or apps like Psensor for CPU temperature. 

The CPU temperature is accessible via `/sys/devices/virtual/thermal/thermal_zone0/temp`. 

`cat /sys/devices/virtual/thermal/thermal_zone0/hwmon1/name`

Result: `acpitz`

This seems to be short for "ACPI Thermal Zone". 

`cat  /sys/devices/virtual/thermal/thermal_zone0/type`

Result: `acpitz`

`cat /sys/devices/virtual/thermal/thermal_zone0/temp`

Result: `48800` (or similar)

Divide by 1000 and if the result is between 20C and 120C or so, and `type` or `hwmon1/name` is `acpitz` it should be a valid CPU temp. Theoretically. 

If you can try accessing `/sys/devices/virtual/thermal/thermal_zone0/temp` if there is no response from psutil on core temps, I think a lot of these older machines could be supported for at least the main CPU temp. 

This machine is a single-core Pentium 4, as many machines prior to Core Duo will be, so there would normally be only one main CPU temp. There were a few late-generation Pentiums with hyper-threading which would probably be reported as dual-core CPUs, but they are a couple of years newer and may be better supported by apps like psutil already.

**Comment by aristocratos** [comment]
@RedBearAK 
The `acpitz` group is reported by `psutil.sensors_temperatures()` if it exists, the problem lies in the sensor names. Unless the group name contains something detectable as a cpu or the sensor itself has a known cpu sensor name I can't really assume it to be a cpu sensor.

In this case group `acpitz` and sensor `temp` or `temp1` could be a cpu proximity sensor, but it's not certain by any means. For example in my system `acpitz` only contains one sensor that is unnamed and is definitively not the cpu temp. And this will depend on the motherboard used, so it's not safe to assume `acpitz` `temp1` would be cpu temp based on the processor used either.

Could you print the output of `python3 -c "import psutil; print(psutil.sensors_temperatures())"`?

Maybe there is some recognizable sensor name in there.

**Comment by RedBearAK** [comment]
As usual, I've misunderstood the true nature of the problem. 

```
{'acpitz': [shwtemp(label='', current=51.8, high=84.8, critical=84.8)]}
```

Soooo... I guess the problem is that the sensor label is blank? 

That's the entire output of the command above. Just that one line.

**Comment by aristocratos** [comment]
@RedBearAK 
I'm guessing if `acpitz` only has one value and the value is named `temp` it is stripped since the named tuple is already named `shwtemp`. But can't really see a way of identifying the sensor, and there would likely be a lot of erroneous cpu temps if I added this value as a cpu temp if no others were detected.

**Comment by aristocratos** [comment]
Actually, it seems like psutil doesn't detect the names from `acpitz` at all possibly, from my system:
`'acpitz': [shwtemp(label='', current=27.8, high=101.0, critical=101.0), shwtemp(label='', current=29.8, high=101.0, critical=101.0)]`

Both sensors are unnamed and neither correspond to my actual cpu temps.

**Comment by RedBearAK** [comment]
Unfortunate. Looks like this may not have a simple resolution. 

What about having an option for the user to input a manual file name to read from? Then it would be up to the user to input the right path. Most machines with this issue would be single-core, so C1 temp and CPU temp would naturally be the same. So reading from a single file source shouldn't normally be a problem. 

It's frustrating that the temperature info is right there, it's just not identifying itself properly. I have gone through CPU stress testing in the past to verify that sensor is definitely the CPU temp on this machine. Its fluctuations correspond directly to CPU activity. 

It's actually a desktop socket 478 CPU that I've managed to upgrade from 1.6GHz to a 2.8GHz chip. You should see the heat sink, it weighs about as much as my iPad Pro. It's quite a ridiculous machine. 

Oh well, it's a nearly 20 year old laptop anyway. It should be going to college or looking for a job.

**Comment by RedBearAK** [comment]
@aristocratos 

So I've been thinking about this and I'd like to ask you to have BpyTOP always show the CPU temps when "Check temp" is "True", even **when there is no temperature data available**. 

I feel like this would be good for a couple of reasons. 

1) It lets the user choose, if they want to, to avoid the formatting issue of not having enough space to see the whole CPU name. Even my MacBook Pro can't show it's whole shortened CPU name when the temps are disabled. The "i7-2860QM" gets the "QM" chopped off. Seven characters. 

2) It actually shows in a positive, understandable way, that the temperature data is unavailable. Which is actually useful information, showing that the app is just having trouble getting any temp data, not that the temperature part of the UI is broken. 

I feel like this would be much better than simply not displaying the temperature bits in the UI at all when there is no temperature data, which just tends to look like a malfunction of the UI. If I set "Check temp" to "True", I should see something being displayed, even if it's just some lines that say "N/A". That's sort of a basic UI rule. If I enable some UI option, I should see it, even if it turns out to show me useless information. 

Then the user can choose whether to keep that on display or set "Check temp" to "False" to make the temp display go away entirely. The display of the temp UI elements becomes entirely controlled by the user's choice, rather than mysteriously going missing. 

Instead of bug reports where someone says, "I don't see any temps," they'll say, "My temps are showing as N/A on this specific platform". A subtle but clarifying difference. 

What do you think? 

I like that there are so many ways that BpyTOP will aut-adapt to circumstances like changing labels when the window size shrinks (brilliant stuff, by the way), but I don't think automatically hiding the temp display should be part of that adaptation. After all, it's not BpyTOP's fault if psutil can't provide any temp data. Leave it up to the user to decide if they want to see that problem and be reminded to investigate why it's happening, or to hide it.

**Comment by aristocratos** [comment]
@RedBearAK 
> What about having an option for the user to input a manual file name to read from?

Reading from a file shouldn't be necessary since psutil should pick up any sensor, but i could probably add an option for "cpu sensor" in the options, where user could pick a sensor from a list of all detected sensors or leave it at "Auto" for auto detection.

> So I've been thinking about this and I'd like to ask you to have BpyTOP always show the CPU temps...

Yeah, can't say I like that idea. It would essentially mean removing functionality (auto hide) and showing unnecessary information (empty temps) just to accommodate low priority information (cpu name) just in case the cpu name would be clipped by the width. And also there already is a solution to the problem; setting a custom cpu name that fits the size in the options.

> Instead of bug reports where someone says, "I don't see any temps," they'll say, "My temps are showing as N/A on this specific platform". A subtle but clarifying difference.

Not sure what you are basing that argument on. So far most reports regarding temperature not showing has included cpu name and often output from lm-sensors. So can't see why people would offer up more information when reporting if it were to say "N/A" as opposed to just actually not being available.

**Comment by aristocratos** [comment]
Cpu temp sensor selection added in v1.0.34

**Comment by RedBearAK** [comment]
@aristocratos 

Version 1.0.35

Regret to report that while I can now select sensor 2/2 "acpitz:1" in the options, no temps appear. Check temp is True. Tried getting out and running the app again. No good. 

Let me know if I can run a different command to help narrow down the issue. Guessing it's just a parsing issue of some kind? 

>Not sure what you are basing that argument on.

I worded that quite badly. I just meant that people would ask why their temps are "N/A" rather than asking why their temps won't display anywhere in the UI. 

Helping the CPU name display would just be a side effect. I'm mainly disagreeing with the idea that seeing ".......... N/A°" is "unnecessary" information. It's information that tells me, especially if I am a new BpyTOP user, that the "Check temp" option is working correctly, but that BpyTOP for some reason can't access the actual temperature data. Which could lead me as a new user to try and look at the options and perhaps discover the "Cpu sensor" option, or some other solution to get the temps working. Or on macOS, to proceed to installing the proper temperature package like coretemp which I might have missed on the initial install. 

If I can't see the temps at all, there is no real way (since we are not the developer) to immediately understand that the temps themselves are broken, because we can't see them. But if you show them as "N/A°" or "ERR°", that immediately says to me, "Oh, the temp display is working but BpyTOP is having trouble accessing the temperature data." 

Outside of this specific example, I am in complete agreement with the philosophy of automatically hiding unimportant elements. That's exactly how I've always tried to design any kind of software, website or script that I've ever been involved with. "Declutter" is my middle name. I just wouldn't do it in this specific case.

**Comment by aristocratos** [comment]
@RedBearAK 
> Regret to report that while I can now select sensor 2/2 "acpitz:1" in the options, no temps appear. 

Should be fixed in v1.0.36

**Comment by RedBearAK** [comment]
@aristocratos 

Works. Very nice! And I have a Pentium 4 now! LOL. 

![temps_fixed](https://user-images.githubusercontent.com/64876997/93815278-2d65c580-fc02-11ea-9a4c-0e44917de65b.png)

One final issue, changing the sensor option doesn't update the display unless the user restarts BpyTOP or manually forces the display to re-render by toggling the "Check temp" option off and on. You can see the temps appear and disappear in the background while the menu is open when you do this, but only if "Background update" is True, of course. 

I might have mistakenly assumed this version still didn't work, but luckily I had the "acpitz:1" sensor selected in the options already, so I saw the temps working right away after the version upgrade, and discovered the view update issue only after playing with the options again. 

So changing the sensor option just needs to trigger the same thing that "Check temp" does, and we should be good. 

==========

Have no idea if this is related in any way (to the 1.0.36 update?), but I just got a very unusual crash when hitting "s" to move the swap from memory to disk. I hit "s" to move from disk to memory, then hit "s" again to move it back and the app crashed with this error in the log. 

```
21/09/20 (11:47:42) | ERROR: Data collection thread failed with exception: '__swap'
Traceback (most recent call last):
  File "/home/kris/.local/lib/python3.6/site-packages/bpytop.py", line 2559, in _runner
    collector._draw()
  File "/home/kris/.local/lib/python3.6/site-packages/bpytop.py", line 2979, in _draw
    MemBox._draw_fg()
  File "/home/kris/.local/lib/python3.6/site-packages/bpytop.py", line 1936, in _draw_fg
    out += f'{Meters.disks_used[name]}{item["used"][:None if big_disk else -2]:>{9 if big_disk else 7}}'
KeyError: '__swap'
21/09/20 (11:47:42) | WARNING: Exiting with errorcode (1). Runtime 0:09:09 
```

Went back in and hit "s" a bunch of times and nothing bad happened. Possibly just a completely random glitch? I have moved the swap many times like this on at least four completely different machines and never seen it crash before. This is also the entire error log on this machine, despite all the time I've had BpyTOP running. It's been very stable for me for a long time now, on all machines I've tried it on. So this was weird.

**Comment by aristocratos** [comment]
@RedBearAK 

> So changing the sensor option just needs to trigger the same thing that "Check temp" does, and we should be good.

Yeah, that was an oversight on my part, will be fixed in next update.

> Have no idea if this is related in any way (to the 1.0.36 update?), but I just got a very unusual crash when hitting "s" to move the swap from memory to disk.

I encountered this crash in early development but only once and never again, I'm pretty sure it happens if you manages to press "s" at the precise microsecond the memory box is being drawn. Will add a check so it waits if any drawing is currently ongoing. There could be other possible bugs like this for other toggles but this is the only one I've encountered myself. I could add "wait if drawing" checks for all toggles but it would increase the latency between keypress to action unnecessarily if the toggle is "safe" from crashing.

**Comment by aristocratos** [comment]
Fixed in v1.0.37

**Comment by RedBearAK** [comment]
@aristocratos 

The updating in v1.0.37 now seems to be working fine on my MacBook. 

But when I fired up the old P4 laptop and tried it, I am now getting a crash when I select the "acpitz:1" sensor, if "Check temp" is True. The error seems oddly similar to the crash when I moved swap. I'll post the error after this because I'm writing this on a different machine. 

==============
I think you've pinpointed the probable cause of the "s" key crash, and the solution. Following the logic I figured it was far more likely to happen when the update interval was reduced to the 100ms minimum, and I was right. I was able to replicate the crash within a few seconds, before I could hit "s" more than a few times. 

Did it again and got a crash within 10 seconds. 

Due to the nature of an app like BpyTOP which is constantly updating the screen, I'd have to say it might be worth at least a moderate amount of time to add the draw check to at least all of the most commonly toggled options. Otherwise you're probably always going to suffer from the "birthday paradox", where an event that you think should be highly improbable ends up happening much more frequently in the real world than you might expect. The shorter the update interval, the more likely the probability of a crash caused by user interaction. 

So an update interval of 100ms is like having a "year" that's only 35 days long instead of 365 days. It becomes highly probable that even a very small group of people randomly chosen will have two people with the same birthday, and then **crash**. 

Obviously this is not an app that will cause any real damage when it crashes. It's up to you how important you think it is for BpyTOP to be as stable as possible. One factor to take into account is that this app looks very nice, so people like keeping it running constantly, and it has a lot of fun options that people tend to like to play with, a **lot**. Together that greatly increases the opportunities for one of these crashes to happen during user interaction.

**Comment by RedBearAK** [comment]
@aristocratos 

I'll post the error below, but strangely I opened a different terminal window to run bpytop while viewing the error log in the first window, and now I can't seem to replicate the crash. 

But there is one remaining glitch which is that if I set the sensor back to "Auto", the temps won't disappear. Which would seem to indicate that setting it from a manual sensor selection back to "Auto" doesn't force a refresh. But the temperature data seems to stop updating, so it seems to actually be set back to "Auto" (which would produce no data on this machine), it just isn't refreshing the display to make the (now bogus) temperatures go away. 

I really can't replicate the crash now. Maybe I had a stuck key or something. 

So v1.0.37 appears to mostly be working on the P4 laptop as well as the MacBook, besides this bug when going back to "Auto" from a manual sensor setting. This would of course be unnoticeable on any machine that has a working CPU sensor under the "Auto" setting. 

Fix that and we should be able to finally give this issue some concrete shoes and send it to sleep with the fishes. I hope. 

```
22/09/20 (21:57:53) | ERROR: Data collection thread failed with exception: 'NotImplementedType' object is not callable
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/bpytop.py", line 2559, in _runner
    collector._draw()
  File "/usr/local/lib/python3.6/dist-packages/bpytop.py", line 2795, in _draw
    CpuBox._draw_fg()
  File "/usr/local/lib/python3.6/dist-packages/bpytop.py", line 1719, in _draw_fg
    out += (f'{THEME.inactive_fg} ⡀⡀⡀⡀⡀{Mv.l(5)}{THEME.gradient["temp"][min_max(cpu.cpu_temp[0][-1], 0, cpu.cpu_temp_crit) * 100 // cpu.cpu_temp_crit]}{Graphs.temps[0](None if cls.resized else cpu.cpu_temp[0][-1])}'
TypeError: 'NotImplementedType' object is not callable
22/09/20 (21:57:54) | WARNING: Exiting with errorcode (1). Runtime 0:00:08 
```

**Comment by aristocratos** [comment]
@RedBearAK 
> Due to the nature of an app like BpyTOP which is constantly updating the screen, I'd have to say it might be worth at least a moderate amount of time to add the draw check to at least all of the most commonly toggled options...

The only toggles able to cause a crash are those that introduce new variables to the draw portion of the function, most of them just toggle a bool value which don't cause any issue. And any toggle in the "proc" box will send an interrupt to the draw function to not spend resources on collection, so they should all be "safe". So I think the toggles should be stable as it is from the checks I've already added.

> But there is one remaining glitch which is that if I set the sensor back to "Auto", the temps won't disappear.

Will be fixed in next update.

> But when I fired up the old P4 laptop and tried it, I am now getting a crash when I select the "acpitz:1" sensor, if "Check temp" is True.

This should only be possible if the collection takes over 1 second (the menus have a 1 second timeout for the collection to finish), which I guess is possible on a Pentium 4. Will be increased in next update.

**Comment by aristocratos** [comment]
Should be fixed in v1.0.38

**Comment by RedBearAK** [comment]
@aristocratos 

> Should be fixed in v1.0.38

LOL. 

I swear I'm not doing this on purpose: 

![auto_temps_vanished](https://user-images.githubusercontent.com/64876997/94351241-2c1e0980-0003-11eb-8320-8a69ea4f1b31.png)
This is going from manual to Auto. ^^^^

![auto_back_to_manual](https://user-images.githubusercontent.com/64876997/94351239-2aecdc80-0003-11eb-892b-7e6f42f1587e.png)
This is going from Auto to manual. ^^^^

The resizing gets fixed in either case if I cycle `Check temp` through True-False-True.

**Comment by RedBearAK** [comment]
@aristocratos 

>The only toggles able to cause a crash are those that introduce new variables to the draw portion of the function, most of them just toggle a bool value which don't cause any issue. And any toggle in the "proc" box will send an interrupt to the draw function to not spend resources on collection, so they should all be "safe". So I think the toggles should be stable as it is from the checks I've already added.

Interesting stuff. Sounds like if you make the swap move function safer that would be pretty much it for crashes caused by general user input. 

I'm still able to cause a crash within a few seconds with the swap move at 100ms update interval. It does become much more difficult at 900ms, but it still happens. Had to hold the "s" key down for a few seconds though, to make it happen in a reasonable amount of time.

**Comment by aristocratos** [comment]
@RedBearAK 
> This is going from manual to Auto.

I forgot a screen redraw after the check, without it the borders won't get redrawn. Will be fixed in next update.

> Had to hold the "s" key down for a few seconds though, to make it happen in a reasonable amount of time.

Should have been enough with the fix already applied, but will add a check in the draw function so any erroneous values get skipped.

**Comment by aristocratos** [comment]
Fixed in v1.0.39

**Comment by RedBearAK** [comment]
@aristocratos 

Confirmed. CPU box updates properly now when moving between Auto and manual sensor, in either direction. 

Did you do something new to mitigate the swap crash? It appears to be more than an order of magnitude better in this version. It's now quite difficult to trigger, even at 100ms. It's not completely gone but if this holds up it seems highly unlikely anyone will ever see the swap crash again under normal usage. 

I think I will close this now since the temp support is wrapped up pretty nicely. Can't find any remaining problems.

**Comment by aristocratos** [comment]
@RedBearAK 
Can you post the error.log from that crash? It must be a different variable causing it if you still are able to crash it.

**Comment by RedBearAK** [comment]
@aristocratos 

Updated to 1.0.40...

Here is the contents of the error log from causing the crash by tapping the "s" key for a few seconds (at 100ms) on Mac: 

```
28/09/20 (09:37:13) | ERROR: Data collection thread failed with exception: 'used'
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/bpytop.py", line 2581, in _runner
    collector._draw()
  File "/usr/local/lib/python3.8/site-packages/bpytop.py", line 3001, in _draw
    MemBox._draw_fg()
  File "/usr/local/lib/python3.8/site-packages/bpytop.py", line 1938, in _draw_fg
    out += (f'{Mv.to(y+cy, x+cx)}{gli}{name.capitalize()[:None if big_mem else 5]+":":<{1 if big_mem else 6.6}}{Mv.to(y+cy, x+cx + cls.mem_width - 3 - (len(mem.swap_string[name])))}{Fx.trans(mem.swap_string[name])}'
KeyError: 'used'
28/09/20 (09:37:13) | WARNING: Exiting with errorcode (1). Runtime 0:00:11 

```


I'll have to fire up the ancient P4 laptop and see if the error it produces is the same. Strangely it seems much easier to replicate the crash on the MacBook Pro today than it was yesterday on the P4 Linux laptop.

**Comment by RedBearAK** [comment]
@aristocratos 

On the P4 laptop it is still quite difficult to replicate the crash. Just sitting and tapping the "s" key seems like it will never cause the crash in a reasonable time, even at 100ms. This seems like new behavior since 1.0.39-ish. However holding down the "s" key will still reliably produce the crash, and very quickly, at least at 100ms.  

```
28/09/20 (09:53:02) | ERROR: Data collection thread failed with exception: 'used'
Traceback (most recent call last):
  File "/home/kris/.local/lib/python3.6/site-packages/bpytop.py", line 2583, in _runner
    collector._draw()
  File "/home/kris/.local/lib/python3.6/site-packages/bpytop.py", line 3003, in _draw
    MemBox._draw_fg()
  File "/home/kris/.local/lib/python3.6/site-packages/bpytop.py", line 1944, in _draw_fg
    out += f'{Mv.to(y+cy, x+cx)}{name.capitalize():{5.5 if cls.mem_size > 1 else 1.1}} {gbg}{Meters.swap[name](None if cls.resized else mem.swap_percent[name])}{mem.swap_string[name][:None if cls.mem_size > 1 else -2]:>{9 if cls.mem_size > 1 else 7}}'; cy += 1 if not cls.graph_height else cls.graph_height
KeyError: 'used'
28/09/20 (09:53:03) | WARNING: Exiting with errorcode (1). Runtime 0:02:19 

```

Seems the "out" line is not the same (probably just platform differences?): 

Mac: 

```
    out += (f'{Mv.to(y+cy, x+cx)}{gli}{name.capitalize()[:None if big_mem else 5]+":":<{1 if big_mem else 6.6}}{Mv.to(y+cy, x+cx + cls.mem_width - 3 - (len(mem.swap_string[name])))}{Fx.trans(mem.swap_string[name])}'
```

Linux: 

```
    out += f'{Mv.to(y+cy, x+cx)}{name.capitalize():{5.5 if cls.mem_size > 1 else 1.1}} {gbg}{Meters.swap[name](None if cls.resized else mem.swap_percent[name])}{mem.swap_string[name][:None if cls.mem_size > 1 else -2]:>{9 if cls.mem_size > 1 else 7}}'; cy += 1 if not cls.graph_height else cls.graph_height
```

Overlayed for easier comparison, Mac on top, Linux below it: 

```
    out += (f'{Mv.to(y+cy, x+cx)}{gli}{name.capitalize()[:None if big_mem else 5]+":":<{1 if big_mem else 6.6}}{Mv.to(y+cy, x+cx + cls.mem_width - 3 - (len(mem.swap_string[name])))}{Fx.trans(mem.swap_string[name])}'
    out += f'{Mv.to(y+cy, x+cx)}{name.capitalize():{5.5 if cls.mem_size > 1 else 1.1}} {gbg}{Meters.swap[name](None if cls.resized else mem.swap_percent[name])}{mem.swap_string[name][:None if cls.mem_size > 1 else -2]:>{9 if cls.mem_size > 1 else 7}}'; cy += 1 if not cls.graph_height else cls.graph_height
```

**Comment by RedBearAK** [comment]
@aristocratos 

Probably irrelevant because it's most likely the same as the other Linux Mint laptop, but here is the crash error from my 2007 MacBook running 64-bit Linux Mint. I am able to replicate the crash by just hitting the "s" key for a while. First error below was while holding down the key, second error was from just tapping the key for about 10 seconds. Oh hey, "Runtime 0:00:11". There you go. 

Seems like I'm able to replicate the crash reliably on wildly different hardware and OS platforms. 2002, 2007, 2011 hardware, 32-bit vs 64-bit Linux, macOS 10.14 Mojave. 

Oh, forgot the Raspberry Pi 4 running Twister OS 1.8.1 (Debian based). And... replicated the crash but had to hold the "s" key down for some time. This is connecting via NoMachine. The error looks the same as the other Linux errors. 

```
28/09/20 (10:51:33) | ERROR: Data collection thread failed with exception: 'used'
Traceback (most recent call last):
  File "/home/kris/.local/lib/python3.8/site-packages/bpytop.py", line 2583, in _runner
    collector._draw()
  File "/home/kris/.local/lib/python3.8/site-packages/bpytop.py", line 3003, in _draw
    MemBox._draw_fg()
  File "/home/kris/.local/lib/python3.8/site-packages/bpytop.py", line 1944, in _draw_fg
    out += f'{Mv.to(y+cy, x+cx)}{name.capitalize():{5.5 if cls.mem_size > 1 else 1.1}} {gbg}{Meters.swap[name](None if cls.resized else mem.swap_percent[name])}{mem.swap_string[name][:None if cls.mem_size > 1 else -2]:>{9 if cls.mem_size > 1 else 7}}'; cy += 1 if not cls.graph_height else cls.graph_height
KeyError: 'used'
28/09/20 (10:51:33) | WARNING: Exiting with errorcode (1). Runtime 0:00:18 

28/09/20 (10:57:25) | ERROR: Data collection thread failed with exception: 'used'
Traceback (most recent call last):
  File "/home/kris/.local/lib/python3.8/site-packages/bpytop.py", line 2583, in _runner
    collector._draw()
  File "/home/kris/.local/lib/python3.8/site-packages/bpytop.py", line 3003, in _draw
    MemBox._draw_fg()
  File "/home/kris/.local/lib/python3.8/site-packages/bpytop.py", line 1944, in _draw_fg
    out += f'{Mv.to(y+cy, x+cx)}{name.capitalize():{5.5 if cls.mem_size > 1 else 1.1}} {gbg}{Meters.swap[name](None if cls.resized else mem.swap_percent[name])}{mem.swap_string[name][:None if cls.mem_size > 1 else -2]:>{9 if cls.mem_size > 1 else 7}}'; cy += 1 if not cls.graph_height else cls.graph_height
KeyError: 'used'
28/09/20 (10:57:25) | WARNING: Exiting with errorcode (1). Runtime 0:00:11 
```

**Comment by aristocratos** [comment]
@RedBearAK 

> ## v1.0.43
> ...
> * Fixed: Additional error handling to prevent crashes from graph and swap toggles

Don't know it it's still possible to force a crash, but haven't been able to do it on my hardware.

**Comment by RedBearAK** [comment]
@aristocratos 

Can still crash v1.0.43 on macOS very quickly and reliably at 100ms by tapping "s" just a few times. 

However at 2000ms I got tired of holding down the "s" key with no results after about a minute. 

At 500ms I tapped for about 30 seconds before getting the crash. Holding the key down shortens the crash time to less than 10 seconds. 

Don't know if this is still the exact same error from before. 

```
18/10/20 (21:47:00) | ERROR: Data collection thread failed with exception: 'used'
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/bpytop.py", line 2601, in _runner
    collector._draw()
  File "/usr/local/lib/python3.8/site-packages/bpytop.py", line 3021, in _draw
    MemBox._draw_fg()
  File "/usr/local/lib/python3.8/site-packages/bpytop.py", line 1957, in _draw_fg
    out += (f'{Mv.to(y+cy, x+cx)}{gli}{name.capitalize()[:None if big_mem else 5]+":":<{1 if big_mem else 6.6}}{Mv.to(y+cy, x+cx + cls.mem_width - 3 - (len(mem.swap_string[name])))}{Fx.trans(mem.swap_string[name])}'
KeyError: 'used'
18/10/20 (21:47:00) | WARNING: Exiting with errorcode (1). Runtime 0:00:07
```

Comparing to earlier reported macOS error. Looks the same: 

```
out += (f'{Mv.to(y+cy, x+cx)}{gli}{name.capitalize()[:None if big_mem else 5]+":":<{1 if big_mem else 6.6}}{Mv.to(y+cy, x+cx + cls.mem_width - 3 - (len(mem.swap_string[name])))}{Fx.trans(mem.swap_string[name])}'
out += (f'{Mv.to(y+cy, x+cx)}{gli}{name.capitalize()[:None if big_mem else 5]+":":<{1 if big_mem else 6.6}}{Mv.to(y+cy, x+cx + cls.mem_width - 3 - (len(mem.swap_string[name])))}{Fx.trans(mem.swap_string[name])}'
```

**Comment by aristocratos** [comment]
@RedBearAK 
Ok, should be fixed in 1.0.43...

Will give you the title of downbringer of systems if you manage to crash it now, at least with the same error!

**Comment by RedBearAK** [comment]
@aristocratos 

I am disappointed by my failure to acquire the title of "Almighty Downbringer of Systems". 

Only tested on macOS so far. 

Set to 100ms update interval. 

Tapped "s" key for a while. 

**No crash.** 

Held "s" key down for at least 60 seconds. 

**No crash.** 

Changing update interval sometimes clobbers half the battery element until something updates it, but no crash. 

Seems you've finally struck at the heart of the beast and robbed me of my lordly title. 

Will confirm on the Linux machines when I have some time.

**Comment by RedBearAK** [comment]
@aristocratos 

Confirming that v1.0.44 appears to fix the swap crash entirely, to the best of my ability to test it. Cannot replicate the crash anymore, even at 100ms, even when holding down "s" key for an extended period, on either of the two Linux Mint laptops I have available, the 2007 MacBook and the old P4 generic laptop. 

Well done indeed.
